parsers/src/main/scala/uk/gov/ons/addressIndex/parsers/Tokens.scala

```diff
@@ -55,11 +56,43 @@ object Tokens extends CrfTokenable {
   override def normalise(tokens: Array[String]): Array[String] = tokens.map(_.toUpperCase)
 
   def tokensToMap(tokens: Seq[CrfTokenResult]): Map[String, String] = {
-    tokens.groupBy(_.label).map {
+    val validatedTokens = validatePostCode(tokens.groupBy(_.label))
+    validatedTokens.map {
       case (token, values) => (token, values.map(_.value).mkString(" "))
     }
   }
 
+  def validatePostCode(tokens: Map[String, Seq[CrfTokenResult]]): Map[String, Seq[CrfTokenResult]] = {
+    val postCodeTokens = tokens.getOrElse(Tokens.postcode, Seq.empty)
+    if(postCodeTokens.size == 1) {
+      val token = postCodeTokens.head.value
+      if(token.length >= 5) {
+        val outcode = token.substring(token.length - 3, token.length)
+        val incode = token.substring(0, token.indexOf(outcode))
+        val regex = "[0-9][A-Z][A-Z]".r
+        val optOutCode = regex.findFirstIn(outcode.toUpperCase)
+        if(optOutCode.isDefined) {
+          tokens.updated(
+            key = Tokens.postcode,
+            value = Seq(
+              CrfTokenResult(
+                value = s"$incode $outcode",
+                label = Tokens.postcode
+              )
+            )
+          )
+        } else {
+          tokens - Tokens.postcode
```

Comment:
My Scala is non existing, so apologies if this is irrelevant. I assume that the - operator here removes the postcode token from the sequence of tokens, correct? If so, then it's correct.